### ISSUE TITLE
PPO implementation does not take care of too large KL Divergence

### ISSUE THREAD

**Issue by ernestum** [hill-a/stable-baselines#213]
While reading the spinningup documentation by OpenAI, I found this interesting note at the end of the ["key equations" section](https://spinningup.openai.com/en/latest/algorithms/ppo.html#key-equations):

> 
> While this kind of clipping goes a long way towards ensuring reasonable policy updates, it is still possible to end up with a new policy which is too far from the old policy, and there are a bunch of tricks used by different PPO implementations to stave this off. In our implementation here, we use a particularly simple method: early stopping. If the mean KL-divergence of the new policy from the old grows beyond a threshold, we stop taking gradient steps.
> 
> When you feel comfortable with the basic math and implementation details, it’s worth checking out other implementations to see how they handle this issue!

So I was wondering what method is used in our PPO2 implementation. I took a look into the source code and could not find anything. Not even early stopping. Even though we DO compute the KL-Divergence during the optimization step. But it never seems to be used.

Either I did overlook something or we should implement at least early stopping to prevent too large updates. It would also be interesting what other implementations do.

**Comment by araffin** [comment]
I think that the first time I see that remark.

>So I was wondering what method is used in our PPO2 implementation.

yes, neither PPO1, nor PPO2 does something like this (this is only present in TRPO where the learning rate is adjusted depending on the kl div).

> Even though we DO compute the KL-Divergence during the optimization step. But it never seems to be used.

Yes, this is only used for logging apparently, but it would not require a lot of changes to include that early stopping. However, this adds a new hyperparameter...

**Comment by Sanderstone** [comment]
Hi,
I am working with PPO2 only for a few weeks, but what is sometimes observe is:
First a model may optimize nicely, but then fails catastrophically. This is often accompanied by a large KL divergence.  
> Yes, this is only used for logging apparently, but it would not require a lot of changes to include that early stopping. However, this adds a new hyperparameter...

I know, more hyperparameters add complexity. Probably early stopping by a KL-limit parameter with an arbitrary high default wouldn't effect anyone negatively but add an option for higher stability. 

Where would you add early stopping in the PPO2 code? It seems in "_train_step" forward-prop, loss-calculation and back-prop are done altogether. In contrast, TRPO has a separate "compute_losses" function. (I use stable baselines 2.4.1)

**Comment by araffin** [comment]
For reference, this was how it was done in the original implementation: https://github.com/joschu/modular_rl/blob/master/modular_rl/ppo.py#L48

**Comment by impeccableaslan** [comment]
Shouldn't we just add these few simple lines of code to ppo2 for now?

```
if (kl > kl_threshold):
      stop training (early stopping)
```

At least until there's a much better method of dealing with too large KL divergence

**Comment by araffin** [comment]
>Shouldn't we just add these few simple lines of code to ppo2 for now?

Yes, this should not be too hard, but I don't really have the time to do it properly for now (add tests + document this feature)

**Comment by smorad** [comment]
@Sanderstone did you find where to make this change?

**Comment by impeccableaslan** [comment]
@smorad Just find where the PPO implementation does gradient descent, then at the line before that check if KL >= KL_minimum, and if true set learning rate of the optimizer to 0 else revert to the original learning rate.

**Comment by araffin** [comment]
done in SB3: https://github.com/DLR-RM/stable-baselines3/pull/419
